stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -0,0 +1,240 @@
+   *
+   * This specialization is parallelized using tbb and works for reverse
+   *   mode autodiff.
+   *
+   * An instance, f, of `ReduceFunction` should have the signature:
+   *   var f(int start, int end, Vec&& vmapped_subset, std::ostream* msgs,
+   * Args&&... args)
+   *
+   * `ReduceFunction` must be default constructible without any arguments
+   *
+   * Each call to `ReduceFunction` is responsible for computing the
+   *   start through end (inclusive) terms of the overall sum. All args are
+   * passed from this function through to the `ReduceFunction` instances.
+   *   However, only the start through end (inclusive) elements of the vmapped
+   * argument are passed to the `ReduceFunction` instances (as the
+   * `vmapped_subset` argument).
+   *
+   * This function distributes computation of the desired sum and the Jacobian
+   * of that sum over multiple threads by coordinating calls to `ReduceFunction`
+   * instances. Results are stored as precomputed varis in the autodiff tree.
+   *
+   * If auto partitioning is true, break work into pieces automatically,
+   *  taking grainsize as a recommended work size (this process
+   *  is not deterministic). If false, break work deterministically
+   *  into pieces smaller than or equal to grainsize. The execution
+   *  order is non-deterministic.
+   *
+   * @param vmapped Sliced arguments used only in some sum terms
+   * @param auto_partitioning Work partitioning style
+   * @param grainsize Suggested grainsize for tbb
+   * @param[in, out] msgs The print stream for warning messages
+   * @param args Shared arguments used in every sum term
+   * @return Summation of all terms
+   */
+  inline var operator()(Vec&& vmapped, bool auto_partitioning, int grainsize,
+                        std::ostream* msgs, Args&&... args) const {
+    const std::size_t num_jobs = vmapped.size();
+
+    if (num_jobs == 0) {
+      return var(0.0);
+    }
+
+    const std::size_t per_job_sliced_terms = count_vars(vmapped[0]);
+    const std::size_t num_sliced_terms = num_jobs * per_job_sliced_terms;
+    const std::size_t num_shared_terms = count_vars(args...);
+
+    vari** varis = ChainableStack::instance_->memalloc_.alloc_array<vari*>(
+        num_sliced_terms + num_shared_terms);
+    double* partials = ChainableStack::instance_->memalloc_.alloc_array<double>(
+        num_sliced_terms + num_shared_terms);
+
+    for (size_t i = 0; i < num_sliced_terms; ++i) {
+      partials[i] = 0.0;
+    }
+    recursive_reducer worker(per_job_sliced_terms, num_shared_terms, partials,
+                             vmapped, msgs, args...);
+
+    if (auto_partitioning) {
+      tbb::parallel_reduce(
+          tbb::blocked_range<std::size_t>(0, num_jobs, grainsize), worker);
+    } else {
+      tbb::simple_partitioner partitioner;
+      tbb::parallel_deterministic_reduce(
+          tbb::blocked_range<std::size_t>(0, num_jobs, grainsize), worker,
+          partitioner);
+    }
```

Comment:
I think this code will fall-over if STAN_THREADS is *not* defined. My suggestion here is to run it like this:

```c++
    auto reduction =
        [&]() {
          if (auto_partitioning) {
            tbb::parallel_reduce(
                tbb::blocked_range<std::size_t>(0, num_jobs, grainsize), worker);
          } else {
            tbb::simple_partitioner partitioner;
            tbb::parallel_deterministic_reduce(
                tbb::blocked_range<std::size_t>(0, num_jobs, grainsize), worker,
                partitioner);
          };
        };

#ifdef STAN_THREADS
    reduction();
#else
    tbb::task_arena single_thread(1);
    single_thread.isolate(reduction);
#endif
```

For this you need to include tbb/task_arena.h

What it does is to let things work as usual. It's just that whenever we do not have a TLS AD tape, we restrict the execution to a single thread (but the prim double only version still runs fully parallel). The isolation ensures that the thread running this will not accept tasks for somewhere else which is necessary in this situation - see https://software.intel.com/en-us/node/506359 .

The alternative is to disable all of this directly in the `reduce_sum` definition in prim where we could do an ifdef like.

```c++
#ifdef STAN_THREADS
    // the thing we have now with parallelism
#else
  // just call the reduce function
  ReduceFunction()(1, size(vamped), ....);
#endif
```

the last thing would be consistent with how we do it for `map_rect` ... so we should probably stick to that.

Opinions? @bbbales2 & @SteveBronder ?

Reply:
I think put the #ifdef in reduce_sum like you suggested. But double check that it indeed falls over. If it doesn't, we don't need to worry.